vshard/util.lua

```diff
@@ -313,6 +313,45 @@ local feature = {
     error_stack = version_is_at_least(2, 4, 0, nil, 0, 0),
 }
 
+local function get_err(instance_type, msg)
+    return (instance_type == "storage" and
+            lerror.vshard(lerror.code.STORAGE_IS_DISABLED, msg) or
+            lerror.vshard(lerror.code.ROUTER_IS_DISABLED, msg))
+end
+
+--
+-- Check if box and storage/router has finished their configuration
+-- and if instance's API wasn't disabled explicitly
+--
+local function check_api_status(instance, is_loaded, instance_type)
```

Comment:
Tbh I would just leave this function inlined in the 2 init.lua files. The reason is that when it is factored out like that, it looks too entangled with magic string constants and members of `instance` object which are not visible nor documented in the function's declaration comment. Even more confusion comes from the necessity to give `is_loaded` explicitly while the other `is_...` are taken from `instance`. I know why it is done like that but still the function looks not too good.

So I would suggest to either
- Inline its usage in init.lua files;
- Or find a way to make it less dependent on `instance` object and its type name. For example, make it accept a single argument `status` which would be `{is_loaded = ..., is_configured = ..., is_enabled = ..., name = ..., errcode = ...}`. The function would check all its stuff like now, update the `status` object, and return `true` or `nil, err` with the given `status.errcode`. `True` means all is fine. `nil, err` means you need to raise the error. And you can fetch updated `is_loaded` from the `status`. Then util.lua won't depend on storage or router neither in names nor in error codes;
- Or factor out only certain parts of the function. For instance `util.box_is_configured`, `util.box_is_running`. And the rest stays in init.lua files. Then code duplication will be insignificant.

Another reason why I would rather recommend to duplicate a few lines than try to extract it in a nice way is that the router shouldn't really depend on `box.cfg` at all. It was a mistake that it calls `box.cfg` internally. And eventually we will most likely drop it along with these `box.info.status` and `type(box.cfg)` checks. Which means extracting them into some common place probably is not needed.

Reply:
In the light of what you have told about removing `box.cfg` from the router in the future inlining seems to be the best solution. After removing there will be only one similar part in `*_api_call_unsafe` concerning `is_configured` and `is_enabled`, which is not very rational to put into common part.